README.md
# practica3

Tercera práctica de la asignatura REDES II. Creación de un cliente de video en tiempo real.
Autores: Alejandro Bravo de la Serna y Miguel González González.

## Utilización del cliente. Preparación
Para usar el cliente a través de la red será necesario habilitar un puerto TCP y un puerto UDP. 
Si se está en un equipo detrás de un NAT, habrá que indicar al router que permita el tráfico TCP y UDP,
cada uno por un puerto a elegir (superior a 1024), y lo reenvíe a la máquina desde la
que se ejecuta el cliente. Asimismo, si se dispone de un firewall en algun punto, deberá permitirse
el paso de los paquetes con esos puertos. Para usar el cliente en local no es necesario, lógicamente.

## Fichero de configuración
Al inicio del programa se carga el fichero config.json con distintos valores usados por el cliente. 

El fichero tiene los siguientes campos:

*  BUFFER_SIZE: Tamaño del buffer de recepción de frames.
*  BUFFER_THRESHOLD: Frames a cargar en el buffer antes de reproducir video.
*  FIXED\_DELAY\_THRESHOLD: Margen mínimo de retardo para tirar paquetes.
*  FPS_REFRESH: Cada cuantos segundos se recalculan los fps.
*  QUALITY_REFRESH: Cada cuantos segundos se recalcula la calidad de compresión.
*  RESOLUTION_REFRESH: Cada cuantos segundos se recalcula la resolución a la que se envía el frame.
*  call_timeout: Tiempo máximo (en segundos) para esperar respuesta en una llamada
*  user_filename: Nombre del fichero del que se carga información del último usuario que uso la aplicación.
*  server_ip: IP del server de descubrimiento.
*  server_port: Puerto en la que se encuentra el servidor de descubrimiento.
*  REPORT_REFRESH: Cada cuanto se envía al otro usuario un report con los paquetes perdidos.
*  REPORT_WEIGHT: Fracción que pondera el valor de los reportes frente a los cáclulos propios. Se tiene que encontrar entre 0 y 1.

Si se nota cierto delay o los fps de la cámara son bajos se recomeinda poner un BUFFER_THRESHOLD menor para reducir el delay.
Si se nota que no llega el vídeo se puede deber a que el FIXED\_DELAY\_THRESHOLD se ha configurado muy bajo.

## Registro del usuario
Al lanzarse la aplicación ejecutando _python3 practica3_client.py_ , el cliente se conecta al servidor de descubrimiento para registrar al usuario.
(En caso de que esta conexión falle (por ejemplo, porque no se disponga de conexión a internet, o el servidor
este caído), la aplicación se cerrará mostrando un mensaje de error).

Cuando la conexión ha sido exitosa, se pedirán los siguientes campos:

*  Nick: Un nombre de usuario.
*  Password: Una clave para iniciar sesión posteriormente.
*  Puerto de control (TCP): El puerto TCP que se ha habilitado
*  Puerto de control (UDP): El puerto UDP que se ha habilitado
*  IP: La IP de internet a través de la que se puede acceder al cliente. 
Si la máquina está detrás de un NAT, por tanto, debe ser la IP del router.
El cliente tratará de autocompletar este campo mediante un servicio externo, pero lo dejará en blanco si no puede.
Para usar el cliente en local, se puede introducir la IP de una de las interfaces de la máquina, o dejarla en blanco.
Si se deja en blanco, se asignará automáticamente la IP de una de las interfaces de la máquina.

Si el nick ya está en el servidor de descubrimiento, la clave debe coincidir con la que se usó para registrarlo por primera vez.
Por tanto, si se obtiene un error de clave incorrecta y es la primera vez que se trata de registrar al usuario, el motivo es que
otra persona ya lo ha registrado.

## Selección de vídeo
El cliente trata de abrir una cámara tras registrarse el usuario. Con los botones de la parte superior del cliente se puede cambiar
entre cámara y fichero de vídeo, aunque cabe mencionar que el cliente ajustará los FPS según la calidad de la transmisión, luego es posible
que el vídeo se capture acelerado o ralentizado, a causa de este ajuste que difiere de los FPS a los que se grabó.

## Realización de llamadas
Con el botón _conectar_ se puede llamar a un usuario introduciendo su nombre. También se muestra una lista con los nombres registrados en el servidor,
para poder seleccionar de ahí. Tras ello, se intentará llamar al usuario y, si este responde, la llamada iniciará automáticamente, mostrando su vídeo.

(_Disclaimer: Este cliente utiliza todos los campos de las cabeceras del protocolo V0 para el ajuste del vídeo, luego si se usa algún otro cliente que no sea este 
y que ignore estas cabeceras o las ajuste erróneamente, es posible que el vídeo entrante no se muestre o lo haga de manera incorrecta._)

## Información de estado
En la barra de estado (parte inferior del cliente) se mostrará información relevante. En la parte izquierda se muestra el estado de la llamada actual.
En la parte central se muestran los parámetros del vídeo que se está enviando. En la parte derecha se muestran los parámetros del vídeo que se está recibiendo,
y la duración de la llamada.

## Control en llamada
Durante una llamada, el usuario puede hacer clic en _Espera_ para poner la llamada en espera, lo que causará que no se envíe ni reciba vídeo. La espera es individual,
es decir, un usuario puede poner la llamada en espera y posteriormente el otro también. Hasta que los 2 usuarios no retiren su espera, no continuará la llamada. En la
parte inferior izquierda (barra de estado), se puede ver en todo momento quién ha puesto la llamada en espera.

Asimismo, se puede finalizar la llamada pulsando el botón de _colgar_ (para realizar otra llamada posterior), o saliendo de la aplicación.

## Pruebas realizadas
Hemos probado el funcionamiento tanto en local como a través de la red entre nosotros y contra clientes de otros compañeros y no hemos detectado ningún problema. También hemos probado con el script _simulate_internet.sh_, 
en local, y se aprecian las pérdidas y retardos por lo que el programa baja la calidad como se espera. Cuando probamos en local con los 2 clientes por loopback con la corrupción de paquetes,
funciona correctamente aunque a veces algunos mensajes TCP que no deberían perderse dan algún fallo, quizás debido a la corrupción de los paquetes de control TCP (no estamos seguros).



CHANGES.md
# Cambios respecto de la entrega temprana

1. Se ha convertido el código para utilizar objetos, atributos y métodos
en lugar de variables globales de estado, como se sugirió.

2. Se ha incluido un fichero de configuración del que leer valores relevantes
para la ejecución del programa. Está explicado en README.md.

3. Se ha introducido e implementado un protocolo V1, que viene explicado en la
wiki del proyecto y permite ajustar la calidad según las pérdidas en el otro
extremo.

4. Se ha actualizado la wiki del proyecto para reflejar estos cambios.

5. Acerca de la función update_screen, nos gustaría comentar que solo se llama
cuando ya se tiene el frame que hay que pintar, y es llamada tanto por el hilo
de recepcion como por el de envío. De esta manera, en cuanto hay disponible
un frame (el que toca sacar del buffer o el grabado por la cámara), se pinta
en la pantalla. Almacenamos también una copia de los últimos frames pintados,
para que si cada hilo va a un ritmo de FPS distinto, se pueda realizar la
actualización de pantalla sin tener que esperar a que el otro obtenga su frame.

video.py
'''
   video.py
   Modulo encargado del apartado criptográfico de securebox
   @author Alejandro Bravo, Miguel Gonzalez
   @version 1.0
   @date 23-04-2020
'''

import threading
import heapq
import time
import cv2
import numpy as np

class VideoBuffer():
    '''Buffer de video: Encapsula el estado y funcionalidades del buffer de video'''

    #Variables globales de estado del modulo
    buffer_lock = threading.Lock() #Cerrojo para el buffer en los 2 hilos (recepcion de la red y extraccion para reproducir)
    buffer_heap = [] #Buffer de video (lista de paquetes)
    buffer_num = 0 #Ultimo paquete que se extrajo del buffer. Esto evita que llegue uno posterior a uno ya emitido.
    timemax = -1 #Retardo fijo. No se reproduciran paquetes pasado este retardo fijo desde su emision.
    packets_lost = [0, 0, 0, 0] #4 contadores de paquetes uno para cada ajuste (calidad,FPS,resolucion,reportes salientes) posible.
    time_last_check_qual = -1 #Timestamp con la ultima vez que se intento actualizar la calidad
    time_last_check_fps = -1 #Timestamp con la ultima vez que se intentaron actualizar los FPS
    time_last_check_res = -1 #Timestamp con la ultima vez que se intento actualizar la resolucion
    buffer_block = True #Indicara al buffer si puede sacar frames o no. Solo lo levantaremos cuando este parcialmente lleno.

    #MEDIDAS PARA AJUSTAR QoS. Cada valor es una fraccion de frames. 
    # Por ejemplo, si desde la ultima comprobacion debieron llegar
    # 10 paquetes, y llegan 6, entonces ha habido una fraccion de 0.4 de perdidas.
    MEDIUM_LOST = 1/15 #Fraccion de frames perdida por segundo que se considera mediocre.
    WORST_LOST = 4/15 #Fraccion de frames perdida por segundo que se considera mala

    #Variables de V1
    last_timestamp = 0 #Marca de tiempo del ultimo reporte
    last_loss_per_second = 0 #Paquetes perdidos por segundo segun el ultimo reporte
    time_last_sent_report = -1 #Timestamp con la ultima vez que se envio reporte de errores
    report_lock = threading.Lock() #Cerrojo para manipular las variables de reportes de perdidas
    using_v1 = False #Indica si se esta usando la version 1, para enviar reportes de perdidas.
    control = None #Modulo de control

    config = None #Objeto de configuracion

    def __init__(self, config):
        '''
        Nombre: __init__
        Descripcion: Constructor que ajusta el objeto de configuracion
        '''
        self.config = config

    def send_frame(self, socket_video, status, frame, numOrden, quality ,resolution, fps):
        '''
        Nombre: send_frame
        Descripcion: Envia un frame comprimido a la ip/puerto indicados.
        Argumentos: socket_video: Socket UDP con el que se envia el frame.
                    status: Contiene el ip y el puerto al que se envia el frame.
                    frame: Frame a enviar.
                    numOrden: Número del frame que se envía.
                    quality: Calidad a la que se comprime.
                    resolution: Resolucion del frame.
                    fps: Numero de frames que se envian por segundo.
        Retorno:
            En caso de que no haya errores devuelve 0
            En caso de error devuelve -1.
        '''
        encimg = compress(frame, quality)
        header = str(numOrden) + "#" + str(time.time()) + "#" + resolution + "#" + str(fps) + "#"
        header = header.encode()
        message = header + encimg
        lengthTot = len(message)

        try:
            lengthSend = socket_video.sendto(message, status)
        except OSError:
            lengthSend = -1
            print("UDP Error: El frame no entra en el datagrama.")

        if(lengthSend != lengthTot):
            return -1
        return 0


    def receive_frame(self, socket_video_rec):
        '''
        Nombre: receive_frame
        Descripcion: Funcion que va recibiendo frames, descomprimiendolos e incluyendolos en un
                    heap.
        Argumentos: socket_video_rec: Socket UDP con el que se recibe el frame.
        Retorno:
            None
        '''
        #Inicializar las variables de control del modulo de video.
        self.buffer_num = -1
        self.timemax = -1
        self.packets_lost = [0, 0, 0, 0]
        self.time_last_check_qual = time.time()
        self.time_last_check_fps = time.time()
        self.time_last_check_res = time.time()
        self.time_last_sent_report = time.time()
        self.last_timestamp = 0

        #Vaciar el socket. Podrian quedar restos de llamadas previas, 
        #cuyos numeros de secuencia lian al contador de paquetes perdidos.
        socket_video_rec.setblocking(0)
        try:
            while socket_video_rec.recvfrom(65535):
                pass
        except OSError:
            #No queda nada que vaciar
            pass
        socket_video_rec.setblocking(1)

        while True:
            data, _ = socket_video_rec.recvfrom(65535)

            if(data == b'END_RECEPTION'):
                return

            with self.buffer_lock:
                video_length = len(self.buffer_heap)

            if(data != None and video_length < self.config.BUFFER_SIZE):
                header,decimg = decompress(data)
                timestamp = float(header[1])
                incoming_fps = int(header[3])

                #Calculo del retardo fijo maximo
                if(self.timemax == -1):
                    #Retardo de red
                    network_delay_estimate = time.time() - timestamp
                    #Retardo que tendra lugar a causa de la espera inicial del buffer
                    buffer_threshold_delay_estimate = self.config.BUFFER_THRESHOLD/incoming_fps
                    #Tiempo maximo de retardo que permitiremos
                    self.timemax = network_delay_estimate + buffer_threshold_delay_estimate + self.config.FIXED_DELAY_THRESHOLD

                #Eliminamos los elementos anteriores al ultimo extraido
                with self.buffer_lock:
                    if((self.buffer_num < int(header[0]))):
                        heapq.heappush(self.buffer_heap, (int(header[0]), header, decimg))
                    #Levantamos el buffer cuando haya un poco de cantidad
                    if(len(self.buffer_heap) > self.config.BUFFER_THRESHOLD):
                        self.buffer_block = False

    def pop_frame(self, quality, fps, resolution, packets_lost_total, min_fps=20, max_fps=40):
        '''
        Nombre: pop_frame
        Descripcion: Extrae un elemento del buffer. Cada elemento es una tripla que
                    contiene el numero de frame, el header y el frame. Ajusta
                    la calidad del video dependiendo de los frames perdidos.
                    heap.
        Argumentos: min_fps : Valor minimo de fps que el QoS puede ajustar.
                    max_fps : Valor maximo de fps que el QoS puede ajustar.

                    Datos que actualiza la funcion (deben pasarse por referencia, envueltos en una lista):
                    quality: Calidad con la que se están comprimiendo los frames
                    fps: Frames que se envian al segundos
                    resolution: Resolucion a la que se captura la imagen
                    self.packets_lost_total: Numero total de paquetes perdidos esta llamada
        Retorno:
            - Si hay elementos: se devuelven 3 elementos. El primero es el numero
            de frame, el segundo elemento es el header y el último el propio frame descomprimido.
            - Si el buffer esta bloqueado se devuelven 3 elementos: -1, una lista vacia
            y un numpy array vacio.
        '''
        #El buffer debe estar desbloqueado, es decir, deben haber llegado suficientes elementos para poder ir extrayendo.
        if(not self.buffer_block):
            with self.buffer_lock:

                #Almacenamos el instante actual
                time_epoch = time.time()

                #Guardamos los fps a los que se envio el frame que vamos a leer
                #buffer[0] -> paquete actual
                #campo 1 -> Header
                #campo 3 del header -> FPS
                fps_entrante = int(self.buffer_heap[0][1][3])
                if fps_entrante <= 0:
                    fps_entrante = 1

                #Si el paquete que toca sacar esta muy retrasado, lo descartamos y sacamos otro
                #buffer[0] -> paquete que toca sacar
                #campo 1 -> Header
                #campo 1 del header -> timestamp
                #Si solo queda un paquete no lo hacemos dado que no hay mas remedio que usar ese
                while len(self.buffer_heap) > 1 and (time_epoch - float(self.buffer_heap[0][1][1]) > self.timemax):
                    heapq.heappop(self.buffer_heap) #Extraccion del paquete descartado

                #Añadimos el numero de paquetes perdidos
                if(self.buffer_num != -1 ):
                    #Paquetes perdidos desde el ultimo que se saco (self.buffer_num)
                    packets_lost_now = self.buffer_heap[0][0] - self.buffer_num -1
                    if(packets_lost_now < 0):
                        packets_lost_now = 0
                    #Se agregan a los 3 contadores (uno para cada parametro de calidad)
                    self.packets_lost[0] += packets_lost_now
                    self.packets_lost[1] += packets_lost_now
                    self.packets_lost[2] += packets_lost_now
                    self.packets_lost[3] += packets_lost_now
                    #Se agregan al conteo total
                    packets_lost_total[0] += packets_lost_now

                #V1: Calculamos fraccion de perdidas segun reporte
                report_fraction = self.last_loss_per_second / fps[0]
                weigth = self.config.REPORT_WEIGHT

                #Ajustamos la calidad de compresión cada QUALITY_REFRESH
                if(time_epoch - self.time_last_check_qual > self.config.QUALITY_REFRESH):

                    quality_fraction = self.packets_lost[0]/(self.config.QUALITY_REFRESH * fps_entrante)
                    if(report_fraction != 0):
                        quality_fraction = quality_fraction * (1-weigth) + report_fraction * weigth

                    if(quality_fraction < self.MEDIUM_LOST):
                        quality[0] = 75
                    elif(quality_fraction < self.WORST_LOST):
                        quality[0] = 50
                    else:
                        quality[0] = 25

                    self.packets_lost[0] = 0
                    self.time_last_check_qual = time_epoch

                #Ajustamos los fps cada FPS_REFRESH
                if(time_epoch - self.time_last_check_fps > self.config.FPS_REFRESH):

                    fps_fraction = self.packets_lost[1]/(self.config.FPS_REFRESH * fps_entrante)
                    if(report_fraction != 0):
                        fps_fraction = fps_fraction * (1-weigth) + report_fraction * weigth

                    if(fps_fraction < self.MEDIUM_LOST):
                        fps[0] = max_fps
                    elif(fps_fraction < self.WORST_LOST):
                        fps[0] = (max_fps + min_fps) // 2
                    else:
                        fps[0] = min_fps

                    self.packets_lost[1] = 0
                    self.time_last_check_fps = time_epoch

                #Ajustamos la resolucion cada RESOLUTION_REFRESH
                if(time_epoch - self.time_last_check_res > self.config.RESOLUTION_REFRESH):

                    resolution_fraction = self.packets_lost[2]/(self.config.RESOLUTION_REFRESH * fps_entrante)
                    if(report_fraction != 0):
                        resolution_fraction = resolution_fraction * (1-weigth) + report_fraction * weigth

                    if(resolution_fraction < self.MEDIUM_LOST):
                        resolution[0] = "640x480"
                    elif(resolution_fraction < self.WORST_LOST):
                        resolution[0] = "320x240"
                    else:
                        resolution[0] = "160x120"

                    self.packets_lost[2] = 0
                    self.time_last_check_res = time_epoch

                #Mandar reportes de perdidas a los que usen V1
                if(self.using_v1):
                    if(time_epoch - self.time_last_sent_report > self.config.REPORT_REFRESH):
                        self.control.send_loss_report(self.packets_lost[3])
                        self.packets_lost[3] = 0
                        self.time_last_sent_report = time_epoch

                #Actualizamos el buffer num al numero del header del primer elemento
                self.buffer_num = self.buffer_heap[0][0]

                #Evitamos vaciado completo en caso de que no se este recibiendo a suficiente ritmo
                if len(self.buffer_heap) == 1:
                    return self.buffer_heap[0]

                return heapq.heappop(self.buffer_heap)

        else:
            return -1, list(), np.array([])

    def set_loss_report(self, lost, timestamp):
        '''
        Nombre: set_lost_report
        Descripcion: Ajusta los datos de perdidas del otro extremo.
        Argumentos:
            lost: paquetes perdidos segun el reporte
            timestamp: Marca de tiempo del reporte
        '''
        with self.report_lock:
            #La primera vez simplemente cogeremos el timestamp.
            if(self.last_timestamp != 0):
                self.last_loss_per_second = lost/(timestamp-self.last_timestamp)
            self.last_timestamp = timestamp

    def set_control(self, control):
        '''
        Nombre: set_control
        Descripcion: Ajusta el modulo de control.
        Argumentos:
            control: Objeto con modulo de control.
        '''
        self.control = control

    def set_using_v1(self):
        '''
        Nombre: set_using_v1
        Descripcion: Indica al modulo de video que el otro cliente usa v1
        '''
        self.using_v1 = True

    def empty_buffer(self):
        '''
        Nombre: empty_buffer
        Descripcion: Vacia el buffer.
        '''
        self.buffer_heap = []
        self.buffer_block = True
        self.last_loss_per_second = 0
        self.using_v1 = False

def compress(frame,quality):
    '''
    Nombre: compress
    Descripcion: Comprime un frame.
    Argumentos: frame: Frame que se va a comprimir.
                quality: String que indica la calidad a la que se va a comprimir.
    Retorno:
        - Si no hay errores se devuelve el frame comprimido.
        - Si hay error al comprimir se devuelve None.
    '''
    # Compresión JPG al 50% de resolución (se puede variar)
    encode_param = [cv2.IMWRITE_JPEG_QUALITY,quality]
    result,encimg = cv2.imencode('.jpg',frame,encode_param)

    if(result == False):
        print('Error al codificar imagen')
        return None

    return encimg.tostring()


def decompress(encimg):
    '''
    Nombre: decompress
    Descripcion: Descomprime un frame.
    Argumentos: frame: Frame que se va a descomprimir.
    Retorno:
        - Si no hay errores se devuelve el header del mensaje y el frame descomprimido.
        - Si hay error al comprimir se devuelve None.
    '''
    count = 0
    ENCODED_HASHTAG = 35 #Codigo de la almohadilla (Si no no funciona)
    for i in range(0, len(encimg)):
        if encimg[i] == ENCODED_HASHTAG:
            count +=1
            if count == 4:
                break

    if(i == len(encimg)):
        print("Error en el datagrama.")
        return None

    header = encimg[:i]
    content = encimg[i+1:]
    header = header.decode()
    header = header.split("#")
    decimg = cv2.imdecode(np.frombuffer(content,np.uint8), 1)

    return header, decimg


discovery.py
'''
    discovery.py
    Modulo que se encarga de la comunicacion con el servidor de descubrimiento.
    @author Miguel Gonzalez, Alejandro Bravo.
    @version 1.0
    @date 22-04-2020
'''

import socket

class Discovery():
    '''Clase de descubrimiento: mantiene el estado y las funcionalidades del modulo de descubrimiento'''

    socket_timeout = 3 #Timeout para la conexion
    server_socket = None #Socket del servidor.
    server_ip = "vega.ii.uam.es" #Parametro con la IP
    server_port = 8000 #Parametro con el puerto

    def __init__(self, server_ip, server_port):
        '''
            Nombre: __init__
            Descripcion: Inicializa la conexion con el servidor
            Argumentos: server_ip : Valor de la IP de discovery.
                        server_port: Valor del puerto de discovery.
            Retorno:
                Lanza una excepcion generica si algo va mal.
        '''
        self.server_ip = server_ip
        self.server_port = server_port

        self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        if self.server_socket == None:
            raise Exception("Error abriendo socket de descubrimiento.")

        print("Conectandose al servidor de descubrimiento...")
        try:
            self.server_socket.settimeout(self.socket_timeout)
            self.server_socket.connect((self.server_ip,self.server_port))
            self.server_socket.settimeout(None)
        except:
            #Timeout
            self.server_socket.close()
            self.server_socket = None
            raise Exception("La conexion al server de descubrimiento ha dado timeout.")

    def server_quit(self):
        '''
            Nombre: server_quit
            Descripcion: Termina la conexion con el servidor, informando al server y limpiando recursos.
            Argumentos:
            Retorno:
                0 si todo ha ido correctamente, -1 en caso de error.
        '''
        if self.server_socket == None:
            print ("Se ha intentado enviar un comando al servidor sin estar conectado a el.")
            return -1
        self.server_socket.send(b'QUIT')
        result = self.server_socket.recv(1024)
        if result.decode() != "BYE":
            print ("Advertencia: El servidor no ha respondido al server_quit. Cerrando conexion...")
        self.server_socket.close()
        print("Desconectado del servidor de descubrimiento.")
        self.server_socket = None
        return 0

    def register_user(self, nickname, password, ip, port, protocols):
        '''
            Nombre: register_user
            Descripcion: Registra un usuario en el sistema.
            Argumentos: nickname: Nombre de usuario a registrar.
                        password: Clave del usuario
                        protocols: Lista de protocolos que soporta
                        ip: Direccion IP desde la que escucha
                        port: Puerto desde el que escucha
            Retorno:
                Timestamp de registro, -1 en caso de clave incorrecta o None en caso de error.
        '''
        if self.server_socket == None:
            print ("Se ha intentado enviar un comando al servidor sin estar conectado a el.")
            return None
        cadena_protocolos = "#".join(protocols) #Creamos la cadena como especifica el protocolo
        mensaje = "REGISTER " + nickname + " " + ip + " " + port + " " + password + " " + cadena_protocolos
        self.server_socket.send(mensaje.encode())
        result = self.server_socket.recv(1024)

        #Obtenemos las palabras de la respuesta
        words = result.decode().split()
        if len(words) < 2:
            #No hay respuesta suficiente
            print ("Error registrando usuario. El servidor no respondio.")
            return None
        elif words[0] == "OK" and words[1] == "WELCOME":
            #Respuesta correcta
            if len(words) < 3:
                #No hay timestamp
                print("Usuario registrado correctamente. El servidor no ha proporcionado timestamp.")
                return "0"
            #Hay timestamp
            print("Usuario registrado correctamente en timestamp: " + words[3])
            return words[2]
        elif words[0] == "NOK" and words[1] == "WRONG_PASS":
            #Respuesta incorrecta
            print("Error registrando usuario. Contraseña incorrecta.")
            return -1
        #Respuesta desconocida
        print("Error desconocido registrando usuario.")
        return None

    def query_user(self,nickname):
        '''
            Nombre: query_user
            Descripcion: Busca la informacion del usuario dado en el sistema.
            Argumentos: nickname: Nombre del usuario.
            Retorno:
                Lista con IP, puerto y lista de protocolos soportados por el cliente, todo como cadenas. None en caso de error.
        '''
        if self.server_socket == None:
            print ("Se ha intentado enviar un comando al servidor sin estar conectado a el.")
            return None
        mensaje = "QUERY " + nickname
        self.server_socket.send(mensaje.encode())
        result = self.server_socket.recv(1024)
        words = result.decode().split()
        if len(words) < 2:
            #No hay respuesta suficiente
            print ("Error buscando usuario. El servidor no respondio.")
            return None
        elif words[0] == "OK" and words[1] == "USER_FOUND":
            #Respuesta correcta
            if len(words) < 6:
                #No hay datos en la respuesta
                print("Error en QUERY: El servidor ha reportado que existe el usuario, pero no ha devuelto sus datos.")
                return -1
            #Hay datos en la respuesta
            print("Usuario " + words[2] + " encontrado correctamente con IP: " + words[3] + ", puerto: " + words[4] + " y protocolos: " + words[5])
            lista_protocolos = words[5].split("#")
            return [words[3],words[4],lista_protocolos]
        elif words[0] == "NOK" and words[1] == "USER_UNKNOWN":
            #Respuesta incorrecta
            print("Error buscando usuario: No existe usuario con ese nombre.")
            return None
        #Respuesta desconocida.
        print("Error desconocido buscando usuario.")
        return None

    def list_users(self):
        '''
            Nombre: list_users
            Descripcion: Muestra un listado de todos los usuarios en el sistema.
            Argumentos:
            Retorno:
                Lista con los datos de cada usuario. Cada dato es [nick ip port ts]. None en caso de error.
        '''
        if self.server_socket == None:
            print ("Se ha intentado enviar un comando al servidor sin estar conectado a el.")
            return None
        mensaje = "LIST_USERS"
        self.server_socket.send(mensaje.encode())

        #Recibir la cantidad adecuada de usuarios
        continue_receiving = True
        result = bytes()
        while continue_receiving:
            result += self.server_socket.recv(1024)
            result_str = result.decode()
            words = result_str.split() #Queremos la tercera palabra que trae cuantos users hay

            #¿Por que este codigo tan raro, en lugar de result_str.count(#)? Pues porque algun
            #gracioso se ha puesto de username: "mua#ja#ja#ja", y el server no lo prohibe.
            #Por tanto contamos como maximo un # por palabra...
            if int(words[2]) <= len([1 for e in words[3:] if '#' in e]):
                continue_receiving = False
                
        if len(words) < 2:
            #No hay respuesta suficiente
            print ("Error listando usuarios. El servidor no respondio.")
            return None
        elif words[0] == "OK" and words[1] == "USERS_LIST":
            #Respuesta correcta
            if len(words) < 3:
                #No hay datos en la respuesta
                print("Advertencia en LIST: El servidor no ha devuelto ningun usuario.")
                return []
            #Obtenemos lista de usuarios.
            lista_usuarios = " ".join(words[3:]).split("#") #A partir de la palabra 3 estan los usuarios
            lista_usuarios = lista_usuarios[:-1] #Eliminamos el espacio en blanco final.

            #Los imprimimos
            i = 1
            print(str(len(lista_usuarios)) + " usuarios encontrados: ")
            for usuario in lista_usuarios:
                print(str(i) + ": " + usuario)
                i += 1

            return list(map(lambda x : x.split() , lista_usuarios))

        elif words[0] == "NOK" and words[1] == "USER_UNKNOWN":
            #Respuesta incorrecta
            print("Error listando usuarios. Servidor da respuesta erronea.")
            return None
        #Respuesta desconocida.
        print("Error desconocido listando usuarios.")
        return None


config.py
'''
   config.py
   Modulo encargado del parseo de valores de configuracion.
   @author Alejandro Bravo, Miguel Gonzalez
   @version 1.0
   @date 16-05-2020
'''

import json

class ConfigParser():
    '''Parser de configuracion: Objeto que almacena los valores de configuracion y puede leerlos de un fichero'''

    # Valores por defecto

    #Parametros constantes del QoS
    BUFFER_SIZE = 256 #Tamano maximo para el buffer.
    BUFFER_THRESHOLD = 10 #Numero de frames que han tenido que llegar para que se reproduzcan frames
    FIXED_DELAY_THRESHOLD = 0.25 #Milisegundos de margen que se permiten como mucho para el retardo fijo.
    FPS_REFRESH = 5.0 #Cada cuanto se intenta reajustar los fps
    QUALITY_REFRESH = 1.0 #Cada cuanto se intenta reajustar la calidad de compresion
    RESOLUTION_REFRESH = 10.0 #Cada cuanto se intenta reajustar la resolucion del video

    #Parametros control
    call_timeout = 15 #Timeout para responder a la llamada
    user_filename = "usuario.json" #Fichero de usuario

    #Config server descubrimiento
    server_ip = "vega.ii.uam.es" #IP del servidor de descubrimiento
    server_port = 8000 #Puerto del servidor de descubrimiento

    #V1
    REPORT_REFRESH = 10.0 #Cada cuanto tiempo se envia un reporte de perdidas
    REPORT_WEIGHT = 0.7 #Peso que se le da a los reportes de perdidas frente a las perdidas de trafico entrante

    #Nombres de las variables que se pueden ajustar
    can_set = ["BUFFER_SIZE", "BUFFER_THRESHOLD", "FIXED_DELAY_THRESHOLD", "FPS_REFRESH", "QUALITY_REFRESH",
               "RESOLUTION_REFRESH", "call_timeout", "user_filename", "server_ip", "server_port", "REPORT_REFRESH", "REPORT_WEIGHT"]

    #Cargamos el fichero
    def __init__(self):
        '''
        Nombre: read_config
        Descripcion: Lee el fichero de configuracion e inicia los valores correspondientes
        En caso de error o parametro faltante, se pondran los valores por defecto.
        '''
        try:
            with open("config.json", "r") as file:
                config = json.load(file)
        except (FileNotFoundError, json.JSONDecodeError):
            print("Error leyendo fichero de configuracion parametros.")
            return

        for key in config:
            if key in self.can_set:
                setattr(self, key, config[key])

        print("CONFIG: Ajustados los parametros: " + str([k for k in config if k in self.can_set]))
        missing = [k for k in self.can_set if k not in config]
        if len(missing) > 0:
            print("CONFIG: Se han ajustado valores por defecto para los parametros faltantes: " + str(missing))
